umbraco/Umbraco-CMS
# Fix media tracking of items added via macro parameters in RTE and Grid

## Details

Fixes #8131 for Umbraco 9

After adding the [item tracking improvements](https://github.com/umbraco/Umbraco-CMS/pull/11919), there was still the issue that media items were not tracked when added via macro parameters in Rich Text Editor and Grid layout. @marcemarc had made a [PR](https://github.com/umbraco/Umbraco-CMS/pull/8614) for Umbraco 8 resolving the issue (🎉 H5YR! 💪 ) and this is the fix for Umbraco 9. 

## Test
- Have the SK installed
  - Create a Partial View Macro File called `InsertFiles.cshtml` ( ❗ _Note: code below_)
  - Allow the macro with the same name to be used in Rich Text Editor and the Grid
  - Add a macro parameter (ex: `Files (files) :  Umbraco.MultipleMediaPicker`)
  - In a **Rich Text area** on a content node, insert the macro and pick a media item as the macro param
    - It should result in the following: 
![image](https://user-images.githubusercontent.com/21998037/158605232-83075143-4d49-4c5f-965b-4f371a63113b.png)

  - Save and Publish
  - View the media item in the backoffice, and you should see that this is now referenced from this page
    - If you pick a folder from the Media section, all its contents will be listed in the macro, but the reference is made to the folder itself and therefore can be found on the folder's info app
  - You can do the same thing with the **Grid**
      - Either pick a Macro directly
      - Or add a RTE in a Grid and reference the macro from there 


---
Contents of `InsertFiles.cshtml`: 
```cshtml
@inherits Umbraco.Cms.Web.Common.Macros.PartialViewMacroPage
@using Umbraco.Cms.Core.Models.PublishedContent
@using Umbraco.Cms.Core
@using Umbraco.Cms.Core.Routing
@using Umbraco.Extensions

@inject IPublishedContentQuery PublishedContentQuery
@inject IVariationContextAccessor VariationContextAccessor
@inject IPublishedUrlProvider PublishedUrlProvider


@{ var mediaIds = Model?.MacroParameters["files"] as string; }

@if (mediaIds != null)
{
    <div class="row">
        @foreach (var mediaId in mediaIds.Split(','))
        {
            var media = PublishedContentQuery.Media(mediaId);

            @* a single image *@
            if (!media.IsDocumentType("Folder"))
            {
                <div class="col-xs-6 col-md-3">
                    <a href="@media.Url(PublishedUrlProvider)" class="thumbnail">@media.Name</a> [@System.IO.Path.GetExtension(@media.Url(PublishedUrlProvider).ToString()).Substring(1)]
                </div>
            }

            @* a folder with images under it *@
            foreach (var image in media.Children(VariationContextAccessor))
            {
                <div class="col-xs-6 col-md-3">
                    <a href="@image.Url(PublishedUrlProvider)" class="thumbnail">@image.Name</a> [@System.IO.Path.GetExtension(@image.Url(PublishedUrlProvider).ToString()).Substring(1)]
                </div>
            }
        }
    </div>
} 
```
